source/opt/common_uniform_elim_pass.cpp

```diff
@@ -130,6 +128,68 @@ ir::Instruction* CommonUniformElimPass::GetPtr(
   return ptrInst;
 }
 
+bool CommonUniformElimPass::IsVolatileStruct(uint32_t type_id) {
+  assert(def_use_mgr_->GetDef(type_id)->opcode() == SpvOpTypeStruct);
+  bool has_volatile_deco = false;
+  dec_mgr_->ForEachDecoration(type_id, SpvDecorationVolatile,
+                              [&has_volatile_deco](const ir::Instruction&){ has_volatile_deco = true;});
+  return has_volatile_deco;
+}
+
+bool CommonUniformElimPass::IsAccessChainToVolatileStructType(const ir::Instruction &AccessChainInst) {
+  assert(AccessChainInst.opcode() == SpvOpAccessChain);
+
+  uint32_t ptr_id = AccessChainInst.GetSingleWordInOperand(0);
+  const ir::Instruction* ptr_inst = def_use_mgr_->GetDef(ptr_id);
+  uint32_t pointee_type_id = GetPointeeTypeId(ptr_inst);
+  const uint32_t num_operands = AccessChainInst.NumOperands();
+
+  // walk the type tree:
+  for (uint32_t idx = 3; idx < num_operands; ++idx) {
+    ir::Instruction* pointee_type = def_use_mgr_->GetDef(pointee_type_id);
+
+    switch (pointee_type->opcode()) {
+      case SpvOpTypeMatrix:
+      case SpvOpTypeVector:
+      case SpvOpTypeArray:
+      case SpvOpTypeRuntimeArray:
+        pointee_type_id = pointee_type->GetSingleWordOperand(1);
+        break;
+      case SpvOpTypeStruct:
+        // check for volatile decorations:
+        if (IsVolatileStruct(pointee_type_id)) return true;
+
+        if (idx < num_operands - 1) {
+          const uint32_t index_id = AccessChainInst.GetSingleWordOperand(idx);
+          const ir::Instruction* index_inst = def_use_mgr_->GetDef(index_id);
+          uint32_t index_value = index_inst->GetSingleWordOperand(2); // TODO: replace with GetUintValueFromConstant()
```

Comment:
There's a universal limit of 16383 struct members.  So this actually should work forever.  :-)
See the table in section 2.17.

It's certainly ok to keep the TODO.